## Part 1
How many pixels stay on after `5` iterations?

### Pretty printing

In [1]:
import pprint


def p_print(data):
    pp = pprint.PrettyPrinter(indent=2, compact=True)
    pp.pprint(data)


def print_img(img):
    print()
    print('\n'.join(img))

### Image transformation methods

In [2]:
def row_to_img(one_row_img):
    return tuple(one_row_img.split('/'))


def rotate(img):
    '''Rotates img clockwise'''
    si, sj, rot = len(img), len(img[0]), []
    for j in range(sj):
        row = []
        for i in range(si):
            row.append(img[si - 1 - i][j])
        rot.append(''.join(row))
    return tuple(rot)


def flip_h(img):
    si, sj, flp = len(img), len(img[0]), []
    for i in range(si):
        row = []
        for j in range(sj):
            row.append(img[i][sj - 1 - j])
        flp.append(''.join(row))
    return tuple(flp)


test, rotated = row_to_img('##./#.#/#../#..'), row_to_img('####/...#/..#.')
flipped_h = row_to_img('.##/#.#/..#/..#')
assert rotated == rotate(test), flipped_h == flip_h(test)


def permutations(img):
    return {
        img,
        flip_h(img),
        rotate(img),
        flip_h(rotate(img)),
        rotate(rotate(img)),
        flip_h(rotate(rotate(img))),
        rotate(rotate(rotate(img))),
        flip_h(rotate(rotate(rotate(img))))
    }


assert len(permutations(test)) == 8  # Must be 8 unique permitations

### Rules parsing

In [3]:
def parse_rules(lines):
    rules = {}
    for line in lines:
        if line.endswith('\n'):
            line = line[:-1]
        in_, out = line.split(' => ')
        in_img = row_to_img(in_)
        out_img = row_to_img(out)
        for in_permutation in permutations(in_img):
            rules[in_permutation] = out_img
    return rules


test_rules = parse_rules(
    '../.# => ##./#../...,.#./..#/### => #..#/..../..../#..#'.split(','))

### Image generation

In [4]:
def split_img(img):
    splits_count = len(img) // 2 if len(img) % 2 == 0 else len(img) // 3
    split_size = len(img) // splits_count
    #print(f'Image size: {len(img)}, splits: {splits_count}, split size: {split_size}')
    splits = []
    for si in range(splits_count):
        splits_row = []
        splits.append(splits_row)
        for sj in range(splits_count):
            split = []
            for i in range(split_size):
                split_row = []
                for j in range(split_size):
                    split_row.append(
                        img[i + split_size * si][j + split_size * sj])
                split.append(''.join(split_row))
            splits_row.append(tuple(split))
    return splits


def join_splits(splits):
    generated = []
    size = len(splits[0][0])
    for splits_row in splits:
        for i in range(size):
            generated.append(''.join([split[i] for split in splits_row]))
    return tuple(generated)


def generate(img, rules):
    processed_splits = []
    for splits_row in split_img(img):
        processed_row = []
        processed_splits.append(processed_row)
        for split in splits_row:
            processed_row.append(rules[split])
    return join_splits(processed_splits)


def generate_n(n, rules):
    img = row_to_img('.#./..#/###')
    for i in range(n):
        #print(f'\nIteration {i+1}:')
        img = generate(img, rules)
        #print(f'\nImage after {i+1} iteration(-s):')
        #print_img(img)
    return img

test_expected = row_to_img('##.##./#..#../....../##.##./#..#../......')
test_actual = generate_n(2, test_rules)
assert test_actual == test_expected

def count_pixels(img):
    return ''.join(img).count('#')

assert count_pixels(test_actual) == 12

### Part 1 image generation

In [5]:
puzzle_rules = parse_rules(list(open('in/day21.txt', 'r')))
puzzle_img = generate_n(5, puzzle_rules)
print('----------\nPart 1 answer: {}'.format(count_pixels(puzzle_img)))

----------
Part 1 answer: 123


### Answers to part 1
`125` is too high.

`123` is the right answer.

## Part 2 - 18 iterations
How many pixels stay on after `18` iterations?

In [6]:
puzzle_img = generate_n(18, puzzle_rules)
print('Part 2 answer: {}'.format(count_pixels(puzzle_img)))

Part 2 answer: 1984683


`1984683` is the correct answer.